Add amcrest binary_sensors

pnbruckner opened the pull request:
## Description:
Add a binary sensor for motion detection to replace the corresponding sensor (which should have been a binary_sensor originally.) Make the default scan interval 5 seconds (rather than the default of 10 seconds for the sensor.) Add a deprecation warning to the motion detection sensor.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9130

## Example entry for `configuration.yaml` (if applicable):
```yaml
amcrest:
  host: !secret amcrest_host
  username: !secret amcrest_username
  password: !secret amcrest_password
  binary_sensors:
    - motion_detected
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

pnbruckner committed:
Add amcrest binary_sensors

Add binary_sensors with option motion_detected. Deprecate motion_detector sensor.

awarecan commented on homeassistant/components/amcrest/__init__.py:
```diff
@@ -157,7 +179,7 @@ def setup(hass, config):
                 }, config)
 
-    return True
+    return len(hass.data[DATA_AMCREST]) >= 1
```
In which case it would return False?

pnbruckner replied:
It would return False if none of the configured cameras worked. If you look at the top of the loop, it tries to communicate with each configured camera. If that fails (AmcrestError), then the camera is not appended to the list. If none work, the list will be empty. In that case setup will return False. Is that ok?

awarecan commented on homeassistant/components/amcrest/binary_sensor.py:
```diff
@@ -0,0 +1,73 @@
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
```
use `async def` instead of `@asyncio.coroutine`

pnbruckner replied:
Good catch. Will do! (It's been so long since I originally wrote this I forgot about that.)

awarecan commented on homeassistant/components/amcrest/binary_sensor.py:
```diff
@@ -0,0 +1,73 @@
+    @property
+    def device_class(self):
+        """Return device class."""
+        return 'motion'
```
Do we have const for this?

pnbruckner replied:
There wasn't when I wrote this, and when I searched for one just a short while ago. However, looks like constants for the binary_sensor device classes were just added 2 days ago by balloob. I can rebase and use it.

awarecan commented on homeassistant/components/amcrest/__init__.py:
```diff
@@ -143,7 +153,19 @@ def setup(hass, config):
+
         if sensors:
+            if 'motion_detector' in sensors:
+                _LOGGER.warning(
```
I will suggest move warning to config schema, see `cv. deprecated()`

pnbruckner replied:
I tried to see how other, existing code handled deprecation, and I did find and look at cv.deprecated(). However, I'm not sure how that could be used here. It looks like it is designed for config keys, and this is not a config key, but rather one of the choices for the list which is a key's data. If you can show me how to use cv.deprecated() here, I'd be happy to do so, but I couldn't see how it would work in this situation.

awarecan replied:
It could be extended to support only validate specific value, but your current implement is okay.

pnbruckner committed:
Update per review

pnbruckner commented:
@awarecan I updated as much as I could per your review, but you'll see I still have some questions for you.

awarecan commented:
Are you planning remove motion_detected  sensor in future release?

MartinHjelmare commented on homeassistant/components/amcrest/__init__.py:
```diff
@@ -94,20 +100,28 @@ def setup(hass, config):
 
     for device in amcrest_cams:
+        name = device[CONF_NAME]
+        if name in hass.data[DATA_AMCREST]:
```
Add a custom validator that checks that all names are unique. See the mysensors component for example of custom validator.

pnbruckner replied:
Will do.

MartinHjelmare commented on homeassistant/components/amcrest/__init__.py:
```diff
@@ -143,7 +153,19 @@ def setup(hass, config):
+                }, config)
+
         if sensors:
+            if 'motion_detector' in sensors:
```
Add a custom validator that checks for deprecation. See mysensors component for example.

pnbruckner replied:
Out of curiosity, why does mysensors have its own implementation of a deprecation validator when one already exists in config_validation.py? Also, the one in mysensors removes the deprecated key. That doesn't seem right. In my understanding, when something is deprecated it means it is still supported but will eventually not be supported. So removing the key would be wrong. It should just be issuing a warning.

I can add a custom deprecation validator, but it will only issue a warning.

MartinHjelmare replied:
Yes, implement it to fit the need here. I mentioned mysensors since it's using similar custom validators.

MartinHjelmare replied:
It the mysensors case, we still "support" the key by removing it, instead of erroring.

The mysensors validator predates the common deprecation validator.

pnbruckner replied:
I did implement it. Did you see the changes?

MartinHjelmare commented on homeassistant/components/amcrest/binary_sensor.py:
```diff
@@ -0,0 +1,73 @@
+            AmcrestBinarySensor(amcrest.name, amcrest.device, sensor_type))
+
+    async_add_devices(amcrest_binary_sensors, True)
+    return True
```
Nothing is checking this return value.

pnbruckner replied:
Ok, I just modeled this after the other platform files (camera.py & sensor.py) that already existed. I'll remove the return statements from them as well.

MartinHjelmare commented on homeassistant/components/amcrest/binary_sensor.py:
```diff
@@ -0,0 +1,73 @@
+
+        _LOGGER.debug('Pulling data from %s binary sensor', self._name)
+
+        if self._sensor_type == 'motion_detected':
```
Remove this check. Device class should be validated in the config schema.

pnbruckner replied:
This isn't the device class. This is the sensor type selection. Granted there's only one binary_sensor type so far, but I modeled this after sensor.py so it would have the same structure should additional binary_sensor types be added in the future. If you think that's a bad idea I can certainly remove it. It could always be added in the future if/when another binary_sensor type is added.

MartinHjelmare replied:
Yeah, let's remove it.

pnbruckner replied:
Ok, will do.

pnbruckner commented:
@awarecan I do plan to remove motion_detector sensor in the future; hence the deprecation warning.

pnbruckner committed:
Update per review

Add custom validators to make sure camera names are unique, and to issue warning if deprecated sensors option motion_detector is used.

async_setup_platform should not return a value.

pnbruckner commented:
@MartinHjelmare I updated per your review as best I could. Let me know if it's ok now or if I still need to do more. Thanks!

pnbruckner committed:
Another review update

Since there is only one type of binary_sensor, remove type test in update method.

MartinHjelmare reviewed (approved):
Looks good! I think we can merge when last comment is addressed and build passes.
